### PR TITLE
Compatibility with Flexibility Hub data model

### PR DESCRIPTION
**Pull request by DavidHerreros**
CryoDRGN protocols have been adapted to work with the new Flexibility Hub data model.

For this branch to work, `scipion-em` must be updated to at least version `3.5.0`.

### CODE REVIEW

**Comment by DavidHerreros** [comment]
> Hi,
> 
> I had a brief look at the flexutils plugin and I think this PR is not good: creating plugin-specific Scipion objects (ParticleFlex etc) and then enforcing them in a different plugin. Why not use the standard SetOfParticles and add attributes for whatever you like (the way it is currently done in the crydrgn plugin)? I'm happy if someone wants to develop their own viewers for cryodrgn outputs (which can be done outside cryodrgn plugin) or save extra attributes to volumes / particles (which is already done in cryodrgn plugin), but I can't accept the proposed PR.

Hi Grigory,

Having extended objects was a good way to handle initially the different parameters CryoDRGN needs. However, we would like CryoDRGN to be able to communicate with other software in what we called the "Flexibility Hub", which is similar to ScipionTomo but for heterogeneity algorithms.

We have seen that each software will need its own extended attributes when working with standard objects, making it hard to handle when we want to use our own developed visualization tools. We would need to consider independently each case, which would be very unpractical.

In addition, extended attributes do not persist in Scipion when doing subsets and other operations, negatively affecting working on real workflows.

Therefore, we decided to define a data model for the Flexibility Hub based on dynamic objects that will persist all this information and simplify the process. Therefore, we would need to adapt CryoDRGN as well, or it might not be able to connect with other programs in the Flexibility Hub.

**Comment by delarosatrevin** [comment]
Hi David, 

I agree with Grigory that this is not a good way to proceed. Imagine now that I create JoseMiguelHub and my own data model. Should I request existing plugins/protocols to change their outputs to fit my needs? It might conflict with other data models if we go that way. I think you need to find a way to deal with that heterogeneity in the Flex plugin but not change the existing plugins, or at least not change in a way that implies a change in their data model. 

Regarding subsets, this might be a general issue if extended properties are not preserved that might be discussed and fixed if necessary.

**Comment by pconesa** [comment]
Hi, just to be clear, flexhub is not just David's plugin...is an attempt to address flexibility methods. In a way it is "defining the model for "flexibility metadata". You are both welcome to participate in its design of course but is not just a plugin of a developer. This is the result of a careful analysis trying to accommodate all flexibility  protocols.

Now, I agree the issue of the properties of the set is an issue, but the fix (I attempted some time ago) did not work and is a tough one.

As far as I know... there it is not possible to "generalize" the flexibility metadata in the same way we have generalized SPA metadata.

The main problem is that Sets partially can't have extended properties thus this suggestion. Otherwise we could extend the "items" instead but unnecessarily.

We can either:

1. go ahead
2. go back to extend set properties (will take some time)
3. Leave cryoDRG out of flex hub.

**Comment by delarosatrevin** [comment]
Hi Pablo, 

The main point was not if was David's plugin or not, most of the plugins/tools are teams' efforts. I just used my name as an example if you have another Hub/Metadata/Logic, we have a problem. I think we should aim for a general solution without breaking existing code or changing core design principles. I think we should keep plugins as close to the original programs/behaviors without adding overhead. 

From your options, I will vote for 2, or if it is too complicated and not worth it the effort, then 3.

**Comment by pconesa** [comment]
Thanks @delarosatrevin , our point is that output of cryodragon and other flexibility analysis tools could benefit from a dedicated metadata model. Still is maybe early for a common stable model, but SPA model is not enough when it comes to make sense of the dinemsionality/umaps etc of the outputs of those methods.
maybe it would be a good idea to present you both what is in our mind that justifies the change? Or maybe after it we see, with your input a better solution with less impact?

File: cryodrgn/protocols/protocol_base.py [hunk 1]
```diff
@@ -36,8 +36,10 @@
 import pyworkflow.utils as pwutils
 from pyworkflow.plugin import Domain
 from pwem.constants import ALIGN_PROJ, ALIGN_NONE
-from pwem.protocols import ProtProcessParticles
-from pwem.objects import SetOfParticles
+from pwem.protocols import ProtProcessParticles, ProtFlexBase
+from pwem.objects import SetOfParticles, ParticleFlex
+
+import cryodrgn.constants as const
```
Comment (azazellochg) [inline cryodrgn/protocols/protocol_base.py:42]: shall we either use this import or "from cryodrgn.constants import" below for constants?

File: cryodrgn/protocols/protocol_base.py [hunk 2]
```diff
@@ -187,12 +189,22 @@ def createOutputStep(self):
...
+        outImgSet = self._createSetOfParticlesFlex(progName=const.CRYODRGN)
         outImgSet.copyInfo(inputSet)
-        outImgSet.copyItems(inputSet, updateItemCallback=self._setZValues,
-                            itemDataIterator=zValues)
-        setattr(outImgSet, WEIGHTS, pwobj.String(self._getFileName('weights_final')))
-        setattr(outImgSet, CONFIG, pwobj.String(self._getFileName('config')))
+        outImgSet.setHasCTF(inputSet.hasCTF())
+        outImgSet.getFlexInfo().setProgName(const.CRYODRGN)
+
+        for particle, zValue in zip(inputSet.iterItems(), zValues):
```
Comment (azazellochg) [inline cryodrgn/protocols/protocol_base.py:197]: should we remove .iterItems() in this case? inputSet is already iterable

File: cryodrgn/protocols/protocol_base.py (hunk 1 above)
Comment (delarosatrevin) [reply cryodrgn/protocols/protocol_base.py:42]: I prefer the way proposed by @azazellochg

File: cryodrgn/protocols/protocol_base.py (hunk 2 above)
Comment (delarosatrevin) [reply cryodrgn/protocols/protocol_base.py:197]: Fewer characters and more concise

**Comment by azazellochg** [comment]
Does this PR depend on https://github.com/scipion-em/scipion-em/pull/433 ?

File: cryodrgn/protocols/protocol_base.py (hunk 1 above)
Comment (DavidHerreros) [reply cryodrgn/protocols/protocol_base.py:42]: Thanks @azazellochg! I have made the changes to reuse "from cryodrgn.constants import" below

File: cryodrgn/protocols/protocol_base.py (hunk 2 above)
Comment (DavidHerreros) [reply cryodrgn/protocols/protocol_base.py:197]: `.iterItems()` removed :)

**Comment by DavidHerreros** [comment]
> Does this PR depend on [scipion-em/scipion-em#433](https://github.com/scipion-em/scipion-em/pull/433) ? I can see [scipion-em/scipion-em#474](https://github.com/scipion-em/scipion-em/pull/474) has been merged

Hi @azazellochg,

Yes, the branch in this PR has been updated according to the changes we agreed some time ago to avoid an extra dependency with `scipion-em-flexutils` plugin. Now the FlexHub data model has been moved to Scipion, so CryoDRGN plugin only needs to import directly from it.

In this way, CryoDRGN will be able to use all the tools in `scipion-em-flexutils` if that plugin is also installed without blocking the possibility to install CryoDRGN alone.

**Comment by azazellochg** [comment]
Hi! So I don't need to wait for https://github.com/scipion-em/scipion-em/pull/433 then. Okay, I will test this branch this weekend and then merge it.

**Comment by DavidHerreros** [comment]
> Hi! So I don't need to wait for https://github.com/scipion-em/scipion-em/pull/433 then. Okay, I will test this branch this weekend and then merge it.

Thank you so much @azazellochg !!

**Review by azazellochg — APPROVED** [review]

**Pull request merged**